### PR TITLE
Feat: claim_fees && claim_ui_fees

### PR DESCRIPTION
**Pull request by gaetbout**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Testing

## What is the current behavior?

Resolves: #291 

## Other information

Updated keys.cairo to use array![] macro everywhere
also had to update some function names from `claim_XXX` to `claimable_XXX`
file `market_store_utils.cairo` wasn't done
no test added

### CODE REVIEW

File: src/fee/fee_utils.cairo [hunk 1]
```diff
@@ -88,7 +90,24 @@ fn claim_fees(
     market: ContractAddress,
     token: ContractAddress,
     receiver: ContractAddress,
-) { // TODO
+) {
+    validate_receiver(receiver);
+
+    let key = keys::claimable_fee_amount_key(market, token);
+
+    let fee_amount = data_store
```
Comment (zarboq) [inline src/fee/fee_utils.cairo:98]: Shouldn't this be retrieved using get_u128 ?

File: src/fee/fee_utils.cairo [hunk 2]
```diff
@@ -88,7 +90,24 @@ fn claim_fees(
...
+        .get_felt252(key)
+        .expect('claim_fees::get_felt252')
+        .try_into()
+        .expect('claim_fees::fee_amount');
+    data_store.set_felt252(key, 0);
```
Comment (zarboq) [inline src/fee/fee_utils.cairo:103]: Same here set_u128

File: src/event/event_emitter.cairo [hunk 3]
```diff
@@ -464,7 +464,7 @@ mod EventEmitter {
         market: ContractAddress,
         receiver: ContractAddress,
         fee_amount: u128,
-        next_pool_value: u128,
+        next_pool_value: felt252,
```
Comment (zarboq) [inline src/event/event_emitter.cairo:467]: why felt252 ?

File: src/fee/fee_utils.cairo [hunk 4]
```diff
@@ -107,6 +126,23 @@ fn claim_ui_fees(
     token: ContractAddress,
     receiver: ContractAddress,
 ) -> u128 {
-    // TODO
-    0
+    validate_receiver(receiver);
+
+    let key = keys::claimable_ui_fee_amount_for_account_key(market, token, ui_fee_receiver);
+    let fee_amount = data_store.get_felt252(key).expect('claim_ui_fees:fee_amount');
```
Comment (zarboq) [inline src/fee/fee_utils.cairo:132]: same here

File: src/fee/fee_utils.cairo [hunk 5]
```diff
@@ -107,6 +126,23 @@ fn claim_ui_fees(
...
+    data_store.set_felt252(key, 0);
+
+    let next_pool_value = data_store
+        .decrement_felt252(keys::claimable_ui_fee_amount_key(market, token), fee_amount)
```
Comment (zarboq) [inline src/fee/fee_utils.cairo:136]: same

File: src/market/market_utils.cairo [hunk 6]
```diff
@@ -51,6 +50,15 @@ struct GetNextFundingAmountPerSizeResult {
     claimable_funding_amount_per_size_delta: PositionType,
 }
 
+struct GetExpectedMinTokenBalanceCache {
+    pool_amount: felt252,
```
Comment (zarboq) [inline src/market/market_utils.cairo:54]: Should use u128 on this struct fields

File: src/market/market_utils.cairo [hunk 7]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+/// # Arguments
+/// * `data_store` - The `DataStore` contract dispatcher.
+/// * `market` - The address of the market
+fn validate_enable_market(data_store: IDataStoreSafeDispatcher, market: Market) {
+    assert(market.market_token.is_non_zero(), 'EmptyMarket');
+    let is_market_disabled = data_store
+        .get_bool(keys::is_market_disabled_key(market.market_token))
+        .expect('validate_enable_market::result')
+        .expect('validate_enable_market::bool');
+    assert(is_market_disabled, 'DisabledMarket');
```
Comment (zarboq) [inline src/market/market_utils.cairo:561]: I think it's assert(!is_market_disabled, 'DisabledMarket');

File: src/market/market_utils.cairo [hunk 8]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+    // use 1 for the getCollateralSum divisor since getCollateralSum does not sum over both the
+    // longToken and shortToken
+    let mut collateral_amount = get_collateral_sum(data_store, market.market_token, token, true, 1);
+    collateral_amount += get_collateral_sum(data_store, market.market_token, token, false, 1);
+    assert(
+        balance_u128 >= collateral_amount,
+        MarketError::INVALID_MARKET_TOKEN_BALANCE_FOR_COLLATERAL_AMOUNT
+    );
+    let claimable_funding_fee_amount = data_store
+        .get_felt252(keys::claimable_funding_amount_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:631]: get_u128

File: src/market/market_utils.cairo [hunk 9]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+/// * `data_store` - The `DataStore` contract dispatcher.
+/// * `market` - The market to increment claimable fees for.
+/// * `token` - The fee token.
+fn get_expected_min_token_balance(
+    data_store: IDataStoreSafeDispatcher, market: Market, token: ContractAddress
+) -> u128 {
+    // get the pool amount directly as MarketUtils.getPoolAmount will divide the amount by 2
+    // for markets with the same long and short token
+    let pool_amount = data_store
+        .get_felt252(keys::pool_amount_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:653]: same

File: src/market/market_utils.cairo [hunk 10]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+        .expect('pool_amount');
+    let swap_impact_pool_amount = get_swap_impact_pool_amount(
+        data_store, market.market_token, token
+    )
+        .into();
+    let claimable_collateral_amount = data_store
+        .get_felt252(keys::claimable_collateral_amount_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:660]: same

File: src/market/market_utils.cairo [hunk 11]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+        .expect('claimable_collateral_amount');
+    let claimable_fee_amount = data_store
+        .get_felt252(keys::claimable_fee_amount_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:663]: same

File: src/market/market_utils.cairo [hunk 12]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+        .expect('claimable_fee_amount');
+    let claimable_ui_fee_amount = data_store
+        .get_felt252(keys::claimable_fee_amount_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:666]: same

File: src/market/market_utils.cairo [hunk 13]
```diff
@@ -538,3 +546,162 @@ fn is_pnl_factor_exceeded(
...
+        .expect('claimable_ui_fee_amount');
+    let affiliate_reward_amount = data_store
+        .get_felt252(keys::affiliate_reward_key(market.market_token, token))
```
Comment (zarboq) [inline src/market/market_utils.cairo:669]: same

**Review by zarboq — CHANGES_REQUESTED** [review]

**Comment by gaetbout** [comment]
That's why I had to do all this clumsy type changing

image

File: src/market/market_utils.cairo (hunk 7 above)
Comment (gaetbout) [reply src/market/market_utils.cairo]: Good catch

**Comment by gaetbout** [comment]
@zarboq 
Should have solved all your remarks

**Pull request merged**